From 46cfe8e7379adf1e22b731dece7e38864b93d4c8 Mon Sep 17 00:00:00 2001
From: Balamurugan Selvarajan <quic_bselvara@quicinc.com>
Date: Wed, 23 Mar 2022 23:39:49 +0530
Subject: [PATCH] ath12k: skip NULL buff segments in RDDM

skip the NULL buff segments during RDDM segment copy.
ATT scripts expects the rddm image segments to have
proper address for the num_seg defined in
ath11k_coredump_seg_info in coredump.

Signed-off-by: Balamurugan Selvarajan <quic_bselvara@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/coredump.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/coredump.c
+++ b/drivers/net/wireless/ath/ath12k/coredump.c
@@ -153,6 +153,7 @@ void ath12k_coredump_download_rddm(struc
 	struct image_info *rddm_img, *fw_img;
 	struct ath12k_dump_segment *segment, *seg_info;
 	int i, rem_seg_cnt = 0, len, num_seg, seg_sz;
+	int skip_count = 0;
 
 	ath12k_mhi_coredump(mhi_ctrl, false);
 
@@ -175,6 +176,10 @@ void ath12k_coredump_download_rddm(struc
 
 	seg_info = segment;
 	for (i = 0; i < fw_img->entries ; i++) {
+		if (!fw_img->mhi_buf[i].buf) {
+			skip_count++;
+			continue;
+		}
 		seg_sz = fw_img->mhi_buf[i].len;
 		seg_info->len = PAGE_ALIGN(seg_sz);
 		seg_info->addr = fw_img->mhi_buf[i].dma_addr;
@@ -184,6 +189,10 @@ void ath12k_coredump_download_rddm(struc
 	}
 
 	for (i = 0; i < rddm_img->entries; i++) {
+		if (!rddm_img->mhi_buf[i].buf) {
+			skip_count++;
+			continue;
+		}
 		seg_sz = rddm_img->mhi_buf[i].len;
 		seg_info->len = PAGE_ALIGN(seg_sz);
 		seg_info->addr = rddm_img->mhi_buf[i].dma_addr;
@@ -196,6 +205,7 @@ void ath12k_coredump_download_rddm(struc
 		if (ab->qmi.target_mem[i].type == CALDB_MEM_REGION_TYPE) {
 			if (ath12k_cold_boot_cal ||
 			    ab->hw_params->cold_boot_calib) {
+				skip_count++;
 				continue;
 			}
 		}
@@ -206,6 +216,7 @@ void ath12k_coredump_download_rddm(struc
 		seg_info++;
 	}
 
+	num_seg = num_seg - skip_count;
 	if(!ab->fw_recovery_support) {
 		ath12k_coredump_seg_info.chip_id = ar_pci->dev_id;
 		ath12k_coredump_seg_info.qrtr_id = ar_pci->ab->qmi.service_ins_id;
